src/Validation/DNSCheckValidation.php

```diff
@@ -36,6 +38,36 @@ public function isValid($email, EmailLexer $emailLexer)
             $host = substr($email, $lastAtPos + 1);
         }
 
+        // Get the domain parts
+        $hostParts = explode('.', $host);
+
+        // Reserved Top Level DNS Names (https://tools.ietf.org/html/rfc2606#section-2),
+        // mDNS and private DNS Namespaces (https://tools.ietf.org/html/rfc6762#appendix-G)
+        $reservedTopLevelDnsNames = [
+            // Reserved Top Level DNS Names
+            'test',
+            'example',
+            'invalid',
+            'localhost',
+
+            // mDNS
+            'local',
+
+            // Private DNS Namespaces
+            'intranet',
+            'internal',
+            'private',
+            'corp',
+            'home',
+            'lan',
+        ];
+
+        // Exclude reserved top level DNS names
+        if (count($hostParts) <= 1 || in_array($hostParts[(count($hostParts) - 1)], $reservedTopLevelDnsNames)) {
```

Comment:
If I understand well, 
`$hostParts[(count($hostParts) - 1)]` extracts the actual domain to check, right?
If so, would you mind extracting this to a local var for clarity?
After that you could go for `isset` instead of transversing the whole list if using the hosts as keys.

Reply:
It is two-fold

1. `count($hostParts) <= 1` to identify if it is a local domain (e.g. blah@localhost or blah@localdomain)
2. `in_array($hostParts[(count($hostParts) - 1)], $reservedTopLevelDnsNames)` is checking the if the tld is in the list of reserved top level dns names (e.g. .test, .example, .invalid etc.)

I think extracting it into a local domain is likely to make it more confusing. Could split it into two checks if you think it is really necessary.

Reply:
I think something got lost from the prvous comments, for this one.
My suggestion then would be something like this, to keep it simple yet more clear:
```php
$isLocalDomain = count($hostParts) <= 1;
$isReservedTopLevel = in_array($hostParts[(count($hostParts) - 1)], $reservedTopLevelDnsNames);
if ($isLocalDomain || isReservedTopLevel) {...}
```
What do you think?

Reply:
Agreed